lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,351 @@
+package com.meshrabiya.lib_nearby.nearby
+
+
+import android.content.Context
+import android.util.Log
+import com.google.android.gms.nearby.Nearby
+import com.google.android.gms.nearby.connection.AdvertisingOptions
+import com.google.android.gms.nearby.connection.ConnectionInfo
+import com.google.android.gms.nearby.connection.ConnectionLifecycleCallback
+import com.google.android.gms.nearby.connection.ConnectionResolution
+import com.google.android.gms.nearby.connection.DiscoveredEndpointInfo
+import com.google.android.gms.nearby.connection.DiscoveryOptions
+import com.google.android.gms.nearby.connection.EndpointDiscoveryCallback
+import com.google.android.gms.nearby.connection.Payload
+import com.google.android.gms.nearby.connection.PayloadCallback
+import com.google.android.gms.nearby.connection.PayloadTransferUpdate
+import com.google.android.gms.nearby.connection.Strategy
+import com.ustadmobile.meshrabiya.ext.ip4AddressToInt
+import com.ustadmobile.meshrabiya.log.MNetLogger
+import com.ustadmobile.meshrabiya.mmcp.MmcpMessage
+import com.ustadmobile.meshrabiya.mmcp.MmcpPing
+import com.ustadmobile.meshrabiya.mmcp.MmcpPong
+import com.ustadmobile.meshrabiya.vnet.VirtualPacket
+import com.ustadmobile.meshrabiya.vnet.VirtualPacketHeader
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import java.net.InetAddress
+import java.nio.ByteBuffer
+import kotlin.random.Random
+
+
+class NearbyVirtualNetwork(
+    context: Context,
+    private val name: String,
+    private val serviceId: String,
+    private val virtualIpAddress: String,
+    private val broadcastAddress: String,
+    private val strategy: Strategy = Strategy.P2P_CLUSTER,
+    private val logger: MNetLogger
+) {
+
+    data class EndpointInfo(
+        val endpointId: String,
+        var status: EndpointStatus = EndpointStatus.DISCONNECTED,
```

Comment:
These need to be val. The idea of flows is that everything is an immutable val, if it is changed, then we should use .copy. Otherwise something that is observing the flow won't see the change.